File: plugins/kraken_method_channel/ios/Classes/KrakenMethodChannel.m
#import "KrakenMethodChannel.h"

@implementation KrakenMethodChannel

static KrakenMethodChannel * _instance;

+(instancetype) sharedMethodChannel {
  return [[self alloc] init];
}


+(instancetype)allocWithZone:(struct _NSZone *)zone{
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _instance = [super allocWithZone:zone];
    });
    return _instance;
}

-(void)onAttach:(KrakenMethodChannelPlugin*)plugin channel:(FlutterMethodChannel*) channel {
  _instance->_channel = channel;
  _instance->_krakenMethodChannelPlugin = plugin;
}

- (void)handleMessageCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if (_instance->_handler != nil) {
    _instance->_handler(call, result);
  }
}

- (void)setMessageHandler:(MessageHandler) handler {
  _instance->_handler = handler;
}

-(void) invokeMethod:(NSString *)method arguments:(nullable id) arguments {
  dispatch_async(dispatch_get_main_queue(), ^{
    [_instance->_channel invokeMethod:method arguments:arguments];
  });
}

- (nonnull id)copyWithZone:(nullable NSZone *)zone {
  return _instance;
}

- (nonnull id)mutableCopyWithZone:(nullable NSZone *)zone {
  return _instance;
}


@end

File: plugins/kraken_method_channel/android/src/main/kotlin/com/taobao/kraken_method_channel/KrakenMethodChannel.java
package com.taobao.kraken_method_channel;

import android.os.Handler;
import android.os.Looper;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;

interface OnKrakenMe {
  // this can be any type of method
  void onGeekEvent();
}

public class KrakenMethodChannel {
  private static KrakenMethodChannel instance;
  private KrakenMethodChannelPlugin plugin;
  private MethodChannel channel;
  private MethodCallHandler handler;

  public static KrakenMethodChannel getInstance() {
    if (instance == null) {
      synchronized (KrakenMethodChannel.class) {
        if (instance == null) {
          instance = new KrakenMethodChannel();
        }
      }
    }
    return instance;
  }

  public void setMessageHandler(MethodCallHandler handler) {
    this.handler = handler;
  }

  public void handleMessageCall(MethodCall call, Result result) {
    if (this.handler != null) {
      this.handler.onMethodCall(call, result);
    }
  }

  public void invokeMethod(final String method, final Object arguments) {
    final KrakenMethodChannel self = this;
    new Handler(Looper.getMainLooper()).post(new Runnable() {
      @Override
      public void run() {
        self.channel.invokeMethod(method, arguments);
      }
    });
  }

  public void onAttach(KrakenMethodChannelPlugin plugin, MethodChannel channel) {
    this.plugin = plugin;
    this.channel = channel;
  }

  public void onDetach() {
    this.plugin = null;
  }
}
